store/src/java/com/zimbra/cs/ml/ClassificationTask.java

```diff
@@ -0,0 +1,158 @@
+package com.zimbra.cs.ml;
+
+import java.util.Set;
+
+import com.google.common.base.Objects;
+import com.google.common.collect.Sets;
+import com.zimbra.common.service.ServiceException;
+import com.zimbra.common.util.ZimbraLog;
+import com.zimbra.cs.ml.callback.ClassificationCallback;
+import com.zimbra.cs.ml.callback.ExclusiveClassCallback;
+import com.zimbra.cs.ml.callback.OverlappingClassCallback;
+import com.zimbra.cs.ml.classifier.Classifier;
+import com.zimbra.cs.ml.classifier.ClassifierManager;
+import com.zimbra.cs.ml.schema.ClassifierInfo;
+
+/**
+ * This class represents a known task to be performed with the classification system.
+ */
+public abstract class ClassificationTask<C extends Classifiable> {
+
+    private String taskName;
+    private ClassificationType taskType;
+    protected ExclusiveClassCallback<C> exclusiveCallback;
+
+    public static enum ClassificationType {
+        BINARY, MULTI_LABEL;
+    }
+
+    public ClassificationTask(String taskName, ClassificationType taskType) {
+        this.taskName = taskName;
+        this.taskType = taskType;
+    }
+
+    public String getTaskName() {
+        return taskName;
+    }
+
+    public ClassificationType getTaskType() {
+        return taskType;
+    }
+
+    public ClassificationTask<C> withExclusiveClassCallback(ExclusiveClassCallback<C> callback) {
+        this.exclusiveCallback = callback;
+        return this;
+    }
+
+    public boolean supportsClassifier(Classifier<C> classifier) {
+        if (exclusiveCallback == null) {
+            return false;
+        }
+        Set<String> classifierExclusiveClasses = Sets.newHashSet(classifier.getInfo().getExclusiveClasses());
+        return classifierExclusiveClasses.containsAll(exclusiveCallback.getExclusiveClasses());
+    }
+
+    protected abstract void checkCanRegister() throws ServiceException;
+
+    public void register() {
+        try {
+            checkCanRegister();
+            ZimbraLog.ml.info("registering classification task %s", taskName);
+            ClassifierManager.getInstance().registerClassificationTask(this);
+        } catch (ServiceException e) {
+            ZimbraLog.ml.error("unable to register classification task %s", taskName, e);
+        }
+    }
+
+    /**
+     * Given a classifier, return the appropriate callback
+     */
+    public ClassificationCallback<C> resolveCallback(Classifier<C> classifier) throws ServiceException {
+        String[] exclusiveClasses = classifier.getInfo().getExclusiveClasses();
```

Comment:
I am missing the logic in this method. It seems to always return the same class member (exclusiveCallback) as a simple getter would (e.g. getExclusiveCallback), but the comment says that it returns "appropriate callback" given a classifier. Is there supposed to be any other logic that chooses which callback to return in addition to null check?

Reply:
Ah, I can see how this is a bit unclear. This method is in the abstract `ClassificationTask` class, which handles the common use case of resolving an exclusive class callback. `BinaryClassificationTask` overrides this method to first try to resolve an overlapping class callback, calling `super.resolveCallback(classifier);` if an appropriate overlapping callback is not found.